src/libraries/System.Collections.Immutable/src/System/Collections/Frozen/FrozenSet.cs

```diff
@@ -129,45 +129,64 @@ private static FrozenSet<T> CreateFromSet<T>(HashSet<T> source)
                 {
                     if (analysis.RightJustifiedSubstring)
                     {
-                        if (analysis.IgnoreCase)
+                        if (analysis.IgnoreCaseForHash)
                         {
                             frozenSet = analysis.AllAsciiIfIgnoreCase
                                 ? new OrdinalStringFrozenSet_RightJustifiedCaseInsensitiveAsciiSubstring(entries, stringComparer, analysis.MinimumLength, analysis.MaximumLengthDiff, analysis.HashIndex, analysis.HashCount)
                                 : new OrdinalStringFrozenSet_RightJustifiedCaseInsensitiveSubstring(entries, stringComparer, analysis.MinimumLength, analysis.MaximumLengthDiff, analysis.HashIndex, analysis.HashCount);
                         }
                         else
                         {
-                            frozenSet = analysis.HashCount == 1
-                                ? new OrdinalStringFrozenSet_RightJustifiedSingleChar(entries, stringComparer, analysis.MinimumLength, analysis.MaximumLengthDiff, analysis.HashIndex)
-                                : new OrdinalStringFrozenSet_RightJustifiedSubstring(entries, stringComparer, analysis.MinimumLength, analysis.MaximumLengthDiff, analysis.HashIndex, analysis.HashCount);
+                            if (analysis.HashCount == 1)
+                            {
+                                frozenSet = analysis.IgnoreCaseForEquals
+                                    ? new OrdinalStringFrozenSet_RightJustifiedSingleCharCaseInsensitive(entries, stringComparer, analysis.MinimumLength, analysis.MaximumLengthDiff, analysis.HashIndex)
+                                    : new OrdinalStringFrozenSet_RightJustifiedSingleChar(entries, stringComparer, analysis.MinimumLength, analysis.MaximumLengthDiff, analysis.HashIndex); ;
+                            }
+                            else
+                            {
+                                frozenSet = analysis.IgnoreCaseForEquals
+                                    ? new OrdinalStringFrozenSet_RightJustifiedSubstringCaseInsensitive(entries, stringComparer, analysis.MinimumLength, analysis.MaximumLengthDiff, analysis.HashIndex, analysis.HashCount)
+                                    : new OrdinalStringFrozenSet_RightJustifiedSubstring(entries, stringComparer, analysis.MinimumLength, analysis.MaximumLengthDiff, analysis.HashIndex, analysis.HashCount);
+                            }
                         }
                     }
                     else
                     {
-                        if (analysis.IgnoreCase)
+                        if (analysis.IgnoreCaseForHash)
                         {
```

Comment:
Am I right to assume that this branch implies `analysis.IgnoreCaseForEquals`? Can you add a `Debug.Assert` checking that?